### ISSUE TITLE
forkJoin should emit empty array or object given empty array or object as input

### ISSUE THREAD

**Issue by aquark** [ReactiveX/rxjs#5209]
I'm reopening https://github.com/ReactiveX/rxjs/issues/2816 because the response wasn't satisfactory.

**RxJS version:** 6.5.2

**Code to reproduce:**

```
forkJoin([]).pipe(
  tap(values => console.log(values)),
).subscribe()
```

**Expected behavior:**

Console logs `[]`

**Actual behavior:**

Nothing is logged.

**Additional information:**

Compare the behavior of `Promise.all`:

```
Promise.all([]).then(values => console.log(values))
// logs []
```

Or `Array.every`:

```
[].every(() => false) // returns true
```

This behavior is important for reasoning inductively and for composability. For example, say I have a list of search terms to expand to search results and then I want to concatenate all the results. The natural way to write this is:

```
forkJoin(searchTerms.map(term => getSearchResults(term))).pipe(
  map(allResults => allResults.flat()),
)
```

You would expect an empty list of search terms to produce an empty list of results, however with current `forkJoin` behavior you need to handle that with a special case:

```
forkJoin(searchTerms.map(term => getSearchResults(term))).pipe(
  map(allResults => allResults.flat()),
  defaultIfEmpty(new Array<SearchResult>()),
)
```

I believe that in every use case where the length of the input to `forkJoin` is unknown at compile time, it's simpler for the consuming code if `forkJoin` returns empty output on empty input, so the consuming code doesn't need to treat empty input as a special case; it can iterate over the output value the same way regardless. This is an example of the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern), where for example it's better for an operation which produces a list to produce an empty list in case there is no output rather than a `null` value which must be handled specially.

In the original issue https://github.com/ReactiveX/rxjs/issues/2816 it was argued that "Forkjoin doesn't consider any other case except all sources emits value, so empty array is also short-curcuit to completion immediately." However in case there are no sources, it is logically true that "all sources emit value", so one would therefore expect forkJoin to emit a value.

This reasoning also applies to `forkJoin({})`, which should emit an empty object `{}`.

**Comment by startswithaj** [comment]
Yeah, this is a recurring problem for us when using `combineLatest`. We've implemented our own wrapper function. 

For details on our use case see: 

https://stackblitz.com/edit/rxjs-playground-royl-vxadk4?file=index.ts

**Comment by cartant** [comment]
`zip` needs the same consideration, too.

**Comment by rraziel** [comment]
@cartant I'm looking at combineLatest and zip for #5066
Has this new behavior been validated? So I know if I can make those changes too

**Comment by benlesh** [comment]
Hi @aquark — thank you for taking the time to open this issue and for sharing your thoughtful feedback with the RxJS project. I’m sorry it took us so long to respond. Given how much time has passed and the lack of recent activity, I’m going to close this for now so the tracker reflects the work we’re likely to act on. If this is still relevant on a current RxJS release, please feel welcome to open a new issue with updated context and a reproduction. Thank you again for your contribution and patience.

— Ben
